### PR TITLE
#857 Support multipart files using InputStream from source file

### PR DESCRIPTION
**Pull request by samridh90**
**Motivation:**

- In certain situations it's useful for the user to be able to provide an `InputStream` to a file instead of the `File` object itself for a multipart file part.
- E.g.: Upload large files accessible over a network protocol such as SFTP to a remote server without copying the file to the local filesystem or loading it fully into memory

**Changes:**

- Introduce new `InputStreamPart`type for use in the `addBodyPart(...)` API
- Implement `InputStreamMultipartPart` to transfer `InputStream `to a Netty `ByteBuf` or `WritableByteChannel`
- _Unfortunately did not find a way to implement `protected long transferContentTo(WritableByteChannel target) throws IOException` for zero-copy uploads._
- Added new tests and updated existing tests to cover scenarios similar to `FilePart`

**Results:**

- It is now possible to provide an `InputStream` as a multipart body part

**Related Issue:**

[#857](https://github.com/AsyncHttpClient/async-http-client/issues/857)

### CODE REVIEW

**Comment by zbum** [comment]
👍

File: client/src/main/java/org/asynchttpclient/request/body/multipart/InputStreamPart.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import java.nio.charset.Charset;
+
+import static org.asynchttpclient.util.Assertions.assertNotNull;
+
+public class InputStreamPart extends FileLikePart {
+
+  private final InputStream inputStream;
+  private final long contentLength;
+
+  public InputStreamPart(String name, InputStream inputStream, long contentLength, String fileName) {
```
Comment (slandelle) [inline client/src/main/java/org/asynchttpclient/request/body/multipart/InputStreamPart.java]: small nit: could you please add a constructir without contenLength that sets the value as `-1`? There's a chance most users won't know the stream length beforehand.

File: client/src/test/java/org/asynchttpclient/request/body/multipart/MultipartUploadTest.java [hunk 2]
```diff
@@ -142,6 +157,28 @@ public void sendEmptyFileZeroCopy() throws Exception {
     sendEmptyFile0(false);
   }
 
+  private void sendEmptyFileInputStream0(boolean disableZeroCopy) throws Exception {
+    File file = getClasspathFile("empty.txt");
```
Comment (slandelle) [inline client/src/test/java/org/asynchttpclient/request/body/multipart/MultipartUploadTest.java]: Could you please use a non empty file?

File: client/src/test/java/org/asynchttpclient/request/body/multipart/MultipartUploadTest.java [hunk 3]
```diff
@@ -142,6 +157,28 @@ public void sendEmptyFileZeroCopy() throws Exception {
...
+    try (AsyncHttpClient c = asyncHttpClient(config().setDisableZeroCopy(disableZeroCopy))) {
+      InputStream inputStream = new BufferedInputStream(new FileInputStream(file));
+      Request r = post("http://localhost" + ":" + port1 + "/upload")
+              .addBodyPart(new InputStreamPart("file", inputStream, file.length(), file.getName(), "text/plain", UTF_8)).build();
```
Comment (slandelle) [inline client/src/test/java/org/asynchttpclient/request/body/multipart/MultipartUploadTest.java]: Could you please test with and without Content-Length prior knowledge?

File: client/src/test/java/org/asynchttpclient/request/body/multipart/MultipartUploadTest.java (hunk 2 above)
Comment (samridh90) [reply client/src/test/java/org/asynchttpclient/request/body/multipart/MultipartUploadTest.java]: This helper method is explicitly to test empty file to ensure we're covering the case from issue https://github.com/AsyncHttpClient/async-http-client/issues/1485. I can add an additional test for non-empty file

Comment (slandelle) [reply client/src/test/java/org/asynchttpclient/request/body/multipart/MultipartUploadTest.java]: Yes please.

Comment (samridh90) [reply client/src/test/java/org/asynchttpclient/request/body/multipart/MultipartUploadTest.java]: Done.

**Comment by samridh90** [comment]
I ran into some trouble with the Netty `HttpObjectEncoder` class when I tried to get zero-copy with -1 contentLength to work. Here's what happens:

- In `org/asynchttpclient/netty/request/body/NettyBodyBody.java#write(final Channel channel, NettyResponseFuture<?> future)` we set `msg = new BodyFileRegion((RandomAccessBody) body);`
- In `io/netty/netty-codec-http/4.1.30.Final/netty-codec-http-4.1.30.Final.jar!/io/netty/handler/codec/http/HttpObjectEncoder.class#encode(ChannelHandlerContext ctx, Object msg, List<Object> out)` this cause us to hit the following block
```java
            case 2:
                if (buf != null) {
                    out.add(buf);
                }

                this.encodeChunkedContent(ctx, msg, contentLength(msg), out);
                break;
```
- Taking a look at the `contentLength(msg)` method, this return -1 when `msg` matches `FileRegion`
```java
private static long contentLength(Object msg) {
        if (msg instanceof HttpContent) {
            return (long)((HttpContent)msg).content().readableBytes();
        } else if (msg instanceof ByteBuf) {
            return (long)((ByteBuf)msg).readableBytes();
        } else if (msg instanceof FileRegion) {
            return ((FileRegion)msg).count();
        } else {
            throw new IllegalStateException("unexpected message type: " + StringUtil.simpleClassName(msg));
        }
    }
```
- In `encodeChunkedContent(ChannelHandlerContext ctx, Object msg, long contentLength, List<Object> out)`:
```java
private void encodeChunkedContent(ChannelHandlerContext ctx, Object msg, long contentLength, List<Object> out) {
        ByteBuf buf;
        if (contentLength > 0L) {
            String lengthHex = Long.toHexString(contentLength);
            buf = ctx.alloc().buffer(lengthHex.length() + 2);
            buf.writeCharSequence(lengthHex, CharsetUtil.US_ASCII);
            ByteBufUtil.writeShortBE(buf, 3338);
            out.add(buf);
            out.add(encodeAndRetain(msg));
            out.add(CRLF_BUF.duplicate());
        }

        if (msg instanceof LastHttpContent) {
            HttpHeaders headers = ((LastHttpContent)msg).trailingHeaders();
            if (headers.isEmpty()) {
                out.add(ZERO_CRLF_CRLF_BUF.duplicate());
            } else {
                buf = ctx.alloc().buffer((int)this.trailersEncodedSizeAccumulator);
                ByteBufUtil.writeMediumBE(buf, 3149066);
                this.encodeHeaders(headers, buf);
                ByteBufUtil.writeShortBE(buf, 3338);
                this.trailersEncodedSizeAccumulator = 0.2F * (float)padSizeForAccumulation(buf.readableBytes()) + 0.8F * this.trailersEncodedSizeAccumulator;
                out.add(buf);
            }
        } else if (contentLength == 0L) {
            out.add(encodeAndRetain(msg));
        }
    }
```
- We basically miss all the blocks since the `encodeChunkedContent` method doesn't handle `contentLength < 0L`.

@slandelle : Not sure if my solution is acceptable, I'd be open to better ideas.

**Comment by slandelle** [comment]
samridh90 I don't think it's worth trying to contribute chunked transfert encoding support for Netty's FileRegion. We can just disable [bypass zero-copy](https://github.com/AsyncHttpClient/async-http-client/blob/master/client/src/main/java/org/asynchttpclient/netty/request/body/NettyBodyBody.java#L56) when content-length < 0.
InputStream isn't zero-copy anyway, such it would only be suboptimal when sending a multipart that mixes a local file and an InputStream.
Then, don't bother implementing zero-copy for InputStreamPart as it shouldn't be used.

WDYT?

**Comment by samridh90** [comment]
@slandelle : Sounds good to me. I'll get rid of the zero-copy for InputStreamPart.

File: client/src/main/java/org/asynchttpclient/netty/request/body/NettyBodyBody.java [hunk 4]
```diff
@@ -54,13 +54,7 @@ public void write(final Channel channel, NettyResponseFuture<?> future) {
 
     Object msg;
     if (body instanceof RandomAccessBody && !ChannelManager.isSslHandlerConfigured(channel.pipeline()) && !config.isDisableZeroCopy()) {
```
Comment (slandelle) [inline client/src/main/java/org/asynchttpclient/netty/request/body/NettyBodyBody.java]: There was some misunderstanding. In my previous comment, I pointed to this line. If you check the content-length here, you can force to the BodyChunkedInput branch if it's < 0. Hence, There's no need to advice the user to disable zero-copy globally, it can be completely transparent.

File: README.md [hunk 5]
```diff
@@ -110,8 +110,11 @@ Use the `addBodyPart` method to add a multipart part to the request.
 This part can be of type:
 * `ByteArrayPart`
 * `FilePart`
+* `InputStreamPart`
 * `StringPart`
 
+**NOTE**: `InputStreamPart` does not support zero-copy transfers. Explicitly disable zero-copy using `config.setDisableZeroCopy(true)` when one of the body parts is an `InputStreamPart`.
```
Comment (slandelle) [inline README.md]: Please revert

Comment (samridh90) [reply README.md]: Thanks for clearing that up.
I've reverted the changes and added a check to BodyChunkedInput if contentLength  0

File: client/src/main/java/org/asynchttpclient/request/body/multipart/InputStreamPart.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
+package org.asynchttpclient.request.body.multipart;
```
Comment (slandelle) [inline client/src/main/java/org/asynchttpclient/request/body/multipart/InputStreamPart.java]: Could you please add the license file header:

```
/*
 * Copyright (c) 2018 AsyncHttpClient Project. All rights reserved.
 *
 * This program is licensed to you under the Apache License Version 2.0,
 * and you may not use this file except in compliance with the Apache License Version 2.0.
 * You may obtain a copy of the Apache License Version 2.0 at
 *     http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the Apache License Version 2.0 is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
 */
```

File: client/src/main/java/org/asynchttpclient/request/body/multipart/part/InputStreamMultipartPart.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
+package org.asynchttpclient.request.body.multipart.part;
```
Comment (slandelle) [inline client/src/main/java/org/asynchttpclient/request/body/multipart/part/InputStreamMultipartPart.java]: missing file header

File: client/src/test/java/org/asynchttpclient/request/body/InputStreamPartLargeFileTest.java [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
```
Comment (slandelle) [inline client/src/test/java/org/asynchttpclient/request/body/InputStreamPartLargeFileTest.java]: Please fix header: Sonatype is no longer involved in this project and new code should be donated to AHC.

**Comment by slandelle** [comment]
On last nit and we'll be good :)

Thanks!

File: client/src/test/java/org/asynchttpclient/request/body/InputStreamPartLargeFileTest.java (hunk 8 above)
Comment (samridh90) [reply client/src/test/java/org/asynchttpclient/request/body/InputStreamPartLargeFileTest.java]: Fixed

File: client/src/main/java/org/asynchttpclient/request/body/multipart/InputStreamPart.java (hunk 6 above)
Comment (samridh90) [reply client/src/main/java/org/asynchttpclient/request/body/multipart/InputStreamPart.java]: Added, thanks for pointing that out.

**Comment by slandelle** [comment]
Merged. Great job, thanks!

**Comment by slandelle** [comment]
2.6.0 should be available on maven central in 15 min.
